src/vonage/ncco_builder/connect_endpoints.py

```diff
@@ -1,49 +1,62 @@
-from pydantic import BaseModel, HttpUrl, AnyUrl, Field, constr
-from typing import Optional, Dict
+from pydantic import BaseModel, HttpUrl, AnyUrl, constr, field_serializer
+from typing import Dict
 from typing_extensions import Literal
 
 
 class ConnectEndpoints:
     class Endpoint(BaseModel):
-        type: str = None
+        type: Literal['phone', 'app', 'websocket', 'sip', 'vbc'] = None
 
     class PhoneEndpoint(Endpoint):
-        type = Field('phone', const=True)
-        number: constr(regex=r'^[1-9]\d{6,14}$')
-        dtmfAnswer: Optional[constr(regex='^[0-9*#p]+$')]
-        onAnswer: Optional[Dict[str, HttpUrl]]
+        type: Literal['phone'] = 'phone'
+
+        number: constr(pattern=r'^[1-9]\d{6,14}$')
+        dtmfAnswer: constr(pattern='^[0-9*#p]+$') = None
+        onAnswer: Dict[str, HttpUrl] = None
+
+        @field_serializer('onAnswer')
+        def serialize_dt(self, oa: Dict[str, HttpUrl], _info):
+            if oa is None:
+                return oa
+
+            return {k: str(v) for k, v in oa.items()}
 
     class AppEndpoint(Endpoint):
-        type = Field('app', const=True)
+        type: Literal['app'] = 'app'
         user: str
 
     class WebsocketEndpoint(Endpoint):
-        type = Field('websocket', const=True)
+        type: Literal['websocket'] = 'websocket'
+
         uri: AnyUrl
         contentType: Literal['audio/l16;rate=16000', 'audio/l16;rate=8000']
-        headers: Optional[dict]
+        headers: dict
+
+        @field_serializer('uri')
+        def serialize_uri(self, uri: AnyUrl, _info):
+            return str(uri)
 
     class SipEndpoint(Endpoint):
-        type = Field('sip', const=True)
+        type: Literal['sip'] = 'sip'
         uri: str
-        headers: Optional[dict]
```

Comment:
Have had a final read through and it looks like this needs an `= None` to mark it optional, just removing the `Optional` type here has made it required. I think there may be more cases like this where removing the `Optional` type has made it a required field 😖

Reply:
Update: have consulted the [pydantic docs](https://docs.pydantic.dev/latest/migration/#required-optional-and-nullable-fields) and Optional has a specific meaning in this case. I think it's okay to swap it out for `= None` here though as when we serialize models we ignore `None` values anyway